osquery/tables/system/windows/programs.cpp

```diff
@@ -88,6 +181,131 @@ void keyEnumPrograms(const std::string& key,
+          // We can extract some information from the registry key
+          // <PackageName>_<Version>_<Architecture>__<PublisherHash>
+          // Example:
+          // MSTeams_25060.205.3499.6849_arm64__8wekyb3d8bbwe
+          // But all proper MSIX packages have an AppxManifest.xml file
+          VLOG(1) << "Cannot find manifest file:'" + regPath + "'";
+          result.clear();
+          continue;
+        }
+
+        // btime "Birth Time"
+        // When a file is first created, btime is set and does not change
+        // ::Warning:: not all filesystems support btime
+        // Older file systems such as ext3 or FAT32 will have this missing
+        WINDOWS_STAT file_stat;
+        auto rtn = platformStat(filePath.c_str(), &file_stat);
+        if (rtn.ok()) {
+          result["install_date"] = formatTimestampToDate(file_stat.btime);
+        }
+
+        std::string line;
+        std::string buffer;
+        while (std::getline(xmlFile, line)) {
+          if (line.find("<Identity") != std::string::npos) {
+            buffer += line;
+
+            // Continue reading lines until the tag is complete (contains "/>")
+            while (buffer.find("/>") == std::string::npos &&
+                   std::getline(xmlFile, line)) {
+              buffer += line;
+            }
+            auto attributes = parseAttributes(buffer);
+            result["name"] = attributes["Name"];
+            result["publisher"] = attributes["Publisher"];
+            result["version"] = attributes["Version"];
+
+            buffer.clear();
+          }
+
+          if (line.find("<Properties") != std::string::npos) {
+            buffer += line;
+
+            // Continue reading lines until the tag is complete (contains
+            // "</Properties>")
+            while (buffer.find("</Properties>") == std::string::npos &&
+                   std::getline(xmlFile, line)) {
+              buffer += line;
+            }
+            auto displayName = extractTagContent(buffer, "DisplayName");
+            auto publisherDisplayName =
+                extractTagContent(buffer, "PublisherDisplayName");
+
+            // "ms-resource:" prefix means that the string is dynamically
+            // generated from a .pri file .pri file is a binary index of all
+            // localized and scaled resources compiled from .resw files or
+            // .resources at build time
+            if (!displayName.empty() &&
+                displayName.find("ms-resource") == std::string::npos) {
+              result["name"] = displayName;
+            }
+            if (!publisherDisplayName.empty() &&
+                publisherDisplayName.find("ms-resource") == std::string::npos) {
+              result["publisher"] = publisherDisplayName;
+            }
+
+            buffer.clear();
+
+            // done parsing the appxmanifest.xml file for the things we need, so
+            // we don't have to continue reading the file
+            break;
+          }
```

Comment:
I know this might change when you move to reading the whole file with readfile, but if it doesn't, a couple of concerns:

- I would prefer to see us using a stringbuffer rather than appending (and re-allocating) a string repeatedly
- The find operation could probably be done on just the last line read rather than the entire buffer
- I think (but maybe I'm misunderstanding) that this assumes the keys are in a specific order. How do we know whether that is a reasonable assumption?